feat: Iceberg read support

scsmithr opened the pull request.

Xuanwo commented:
Hi, it's great to see that glaredb now supports iceberg! I wanted to let you know that there is an ongoing effort to implement a rust-native version of iceberg upstream at https://github.com/apache/iceberg-rust. You might find it interesting.

scsmithr commented:
> Hi, it's great to see that glaredb now supports iceberg! I wanted to let you know that there is an ongoing effort to implement a rust-native version of iceberg upstream at https://github.com/apache/iceberg-rust. You might find it interesting.

Will definitely check it out (getting late here, but for sure tomorrow). I'd love see if there's stuff we could contribute.

Xuanwo commented:
> I'd love see if there's stuff we could contribute.

That's great news!

scsmithr commented:
PR is getting pretty large, but it's in a good spot to review and merge. About half the lines added are from testdata.

A follow up issue (that I'll open tomorrow) will list out further improvements, including:
- S3, GCS tests
- Get statistics during planning
- Better handling of partitioned tables
- Figure out v1/v2 compatibility. I think we're mostly v2 compatible right now.
- See what could mergeable into iceberg-rust
- And more!

vrongmeal commented on out.parquet:
Did you mean to commit this?

scsmithr replied:
nope

vrongmeal reviewed (commented):
Just a couple of minor comments.

How much time does it take to generate testdata files? If we can generate it easily, should we just commit the script?

scsmithr commented on crates/datasources/src/common/url.rs:
```diff
@@ -23,14 +23,14 @@ impl Display for DatasourceUrlScheme {
         match self {
             Self::File => write!(f, "file"),
-            Self::Http => write!(f, "http(s)"),
+            Self::Http => write!(f, "https"),
```
Renamed, also renamed the relevant method on `DatasourceUrl`, and added a `scheme` method that returns the string.

scsmithr commented:
> How much time does it take to generate testdata files? If we can generate it easily, should we just commit the script?

It's pretty quick, and I did check in the python script I used to generate them. However I wouldn't want to generate the files in CI because that would mean needing to set up python with pyspark and getting the iceberg jar.

universalmind303 commented on testdata/iceberg/tables/lineitem_partitioned/metadata/v1.metadata.json:
```diff
@@ -0,0 +1,148 @@
+{
+  "format-version" : 2,
+  "table-uuid" : "d3224394-5406-4ba2-8abf-8c794a200c89",
+  "location" : "/Users/sean/Code/github.com/glaredb/glaredb/testdata/iceberg/tables/lineitem_partitioned",
```
i think we may need to autogenerate these. the `generatd_iceberg` script fails locally as this is hardcoded to your filesystem.

scsmithr replied:
I accidentally used an expanded path in the script, pushing up a fix.

scsmithr commented on testdata/generate_iceberg.py:
```diff
@@ -0,0 +1,54 @@
+conf.set("spark.sql.catalog.iceberg_catalog.warehouse", OUTPUT_DIR)
+conf.set("spark.sql.parquet.outputTimestampType", "TIMESTAMP_MICROS")
+conf.set("spark.driver.memory", "10g")
+conf.set("spark.jars", f"{SCRIPT_DIR}/iceberg-spark-runtime-3.4_2.12-1.3.0.jar")
```
I don't want to deal with spark anymore

universalmind303 replied:
lmao fair enough.

scsmithr commented on crates/datasources/Cargo.toml:
```diff
@@ -6,6 +6,7 @@ edition = {workspace = true}
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+apache-avro = "0.15"
```
It exports a wrapper around an avro reader for reading into record batches. Datafusion is also on 0.14 right now.

Using the avro stuff directly is scoped to reading the metadata, and allows us to use serde on the types. I don't think there will be a chance for conflict/confusion since avro isn't used outside the `spec` module (which itself doesn't depend on any datafusion stuff).

scsmithr commented:
- Fixed metadata locations in test files (technically absolute paths always work for _us_ but not when running spark).
- Added tests for getting relative paths from metadata paths.
- Automatically find spark jar in current directory in generate iceberg table script.

universalmind303 commented:
one last comment. the `generate_iceberg` script seem to be non-deterministic in terms of file names. When I run it, it creates 50 new files in `testdata/iceberg/tables/*`. 


![image](https://github.com/GlareDB/glaredb/assets/21327470/133d748d-c642-4acf-a1ac-3b6541a194ba)

scsmithr commented:
Follow up items: https://github.com/GlareDB/glaredb/issues/1448
